src/visitors/cvode_visitor.cpp

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2023 Blue Brain Project, EPFL.
+ * See the top-level LICENSE file for details.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "visitors/cvode_visitor.hpp"
+
+#include "ast/all.hpp"
+#include "lexer/token_mapping.hpp"
+#include "pybind/pyembed.hpp"
+#include "utils/logger.hpp"
+#include "visitors/visitor_utils.hpp"
+#include <optional>
+#include <utility>
+
+namespace pywrap = nmodl::pybind_wrappers;
+
+namespace nmodl {
+namespace visitor {
+
+static int get_index(const ast::IndexedName& node) {
+    return std::stoi(to_nmodl(node.get_length()));
+}
+
+static void remove_conserve_statements(ast::StatementBlock& node) {
+    auto conserve_equations = collect_nodes(node, {ast::AstNodeType::CONSERVE});
+    if (!conserve_equations.empty()) {
+        std::unordered_set<ast::Statement*> eqs;
+        for (const auto& item: conserve_equations) {
+            eqs.insert(std::dynamic_pointer_cast<ast::Statement>(item).get());
+        }
+        node.erase_statement(eqs);
+    }
+}
+
+static std::pair<std::string, std::optional<int>> parse_independent_var(
+    std::shared_ptr<ast::Identifier> node) {
+    auto variable = std::make_pair(node->get_node_name(), std::optional<int>());
+    if (node->is_indexed_name()) {
+        variable.second = std::optional<int>(
+            get_index(*std::dynamic_pointer_cast<const ast::IndexedName>(node)));
+    }
+    return variable;
+}
+
+static std::unordered_map<std::string, int> get_name_map(const ast::Expression& node,
+                                                         const std::string& name) {
+    std::unordered_map<std::string, int> name_map;
+    // all of the "reserved" symbols
+    auto reserved_symbols = get_external_functions();
+    // all indexed vars
+    auto indexed_vars = collect_nodes(node, {ast::AstNodeType::INDEXED_NAME});
+    for (const auto& var: indexed_vars) {
+        if (!name_map.count(var->get_node_name()) && var->get_node_name() != name &&
+            std::none_of(reserved_symbols.begin(), reserved_symbols.end(), [&var](const auto item) {
+                return var->get_node_name() == item;
+            })) {
```

Comment:
Please be explicit about converting integers to booleans: e.g. `count > 0`. I think this line is long enough to allow breaking up, e.g. give the third clause a name please.